### PR TITLE
[JIT] clear GraphFunction.optimized_graphs_ after freezing a module

### PR DESCRIPTION
**Pull request by davidberard98**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #68316

Consider the following:
```
class Mod(nn.Module):
    def __init__(self, val):
	super().__init__()
	self.param = nn.Parameter(val)

    def forward(self, x):
	# this method will change during freezing
	return x + self.param

    @torch.jit.export
    def make_prediction(self, x):
	y = x + x
	return self.forward(y)

param = torch.rand([2, 2])

unscripted_mod = Mod(param)
mod = torch.jit.script(unscripted_mod)
mod.eval()
mod = torch.jit.freeze(mod, preserved_attrs=["make_prediction"])`
```

During freezing the following will occur:
1. do some pre-freezing, including inlining; in particular, forward will be inlined into make_prediction. During inlining, forward.optimized_graph() is called, and the result is cached
2. freeze some methods. While freezing forward, the graph associated with the function will get updated. The cached optimized_graphs_ are not updated.

Previously, a call to `mod.forward(x)` would return an exectutor that would run on the old cached optimized_graph(). This would mean that the freezing optimizations would not apply, and potentially that the execution would fail because of parameters removed from the module.

This change clears the optimized_graphs_ cache after running freezing to prevent executing an old version of the graph.

Differential Revision: [D32410862](https://our.internmc.facebook.com/intern/diff/D32410862)

### CODE REVIEW

**Comment by davidberard98** [comment]
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32410862).

File: torch/csrc/jit/api/function_impl.h [hunk 1]
```diff
@@ -100,6 +100,12 @@ struct TORCH_API GraphFunction : public Function {
     return true;
   }
 
+  void clear_optimized_graphs() {
```
Comment (eellison) [inline torch/csrc/jit/api/function_impl.h:103]: is this the right way of clearing out an array ? im not that well-versed on this c++ idiom cc @jjsjann123

**Review by eellison — APPROVED** [review]
😍 😍 😍

File: torch/csrc/jit/api/function_impl.h (hunk 1 above)
Comment (davidberard98) [reply torch/csrc/jit/api/function_impl.h:103]: I suppose std::fill might be better?

Comment (jjsjann123) [reply torch/csrc/jit/api/function_impl.h:103]: std::array::fill. So I suppose something like
`optimized_graphs_.fill(c10::nullopt);`

**Comment by davidberard98** [comment]
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32410862).

**Review by jjsjann123 — APPROVED** [review]

**Pull request closed without merging**